### ISSUE TITLE
Activity Classifier random_split_by_session should support datasets with fewer than 100 sessions

### ISSUE THREAD

**Issue by tetreault** [apple/turicreate#964]
Hello, making this new thread since i erroneously discussed this in an unrelated thread (https://github.com/apple/turicreate/issues/813)

First post: 

Total machine learning noob here but came across Turi because I have a task that requires somewhat reliably going from standing-to-kneeling posture (and ideally vice versa). I was following along with the documentation, and made sure my preprocessed CSV mirrors the format used in the table in the "advanced usage" section. 

I'm running the following python script to generate a model: 

```
import turicreate as tc
dataSFrame = tc.SFrame('motion_data_all.csv')

# Train/test split by recording sessions
train, test = tc.activity_classifier.util.random_split_by_session(dataSFrame, session_id='exp_id', fraction=0.8)

# Create an activity classifier
motion_model = tc.activity_classifier.create(train, session_id='exp_id', target='motion', prediction_window=50)

metrics = motion_model.evaluate(test)
print(metrics['accuracy'])

# Save the model for later use in Turi Create
motion_model.save('motion_model.model')

# Export for use in Core ML
motion_model.export_coreml('MotionActivityClassifier.mlmodel')

# try out the model 
model_test_data = dataSFrame[(dataSFrame['motion'] == 'sit') & (dataSFrame['exp_id'] == 1)][500:1000]
motion_model.predict(model_test_data, output_frequency='per_window')
```

My training data is extremely small, as far as ML training data goes. Approx 3.5k rows of standing/sitting/kneeling, each accurately labeled. In the doc you use the HAPT data and then process it into a format with `exp_id` and the activity label. My data set mirrors that end result. 

When I run the following line in spyder: 

```
metrics = motion_model.evaluate(test)
print(metrics['accuracy'])
```

**I get the following error: `ToolkitError: Size of prediction probability vector(3) != number of classes(1).`**

Just to be clear I'm not completely messing up from the start, I've included some screenshots from the file I use with the following: `dataSFrame = tc.SFrame('motion_data_all.csv')`.

<img width="755" alt="screenshot 2018-08-11 18 24 19" src="https://user-images.githubusercontent.com/5934354/43996590-278bee60-9d94-11e8-8e00-3c38a1cb9838.png">
<img width="764" alt="screenshot 2018-08-11 18 24 34" src="https://user-images.githubusercontent.com/5934354/43996592-2de67262-9d94-11e8-9188-acac9ca507fa.png">

When I run the line: `motion_model = tc.activity_classifier.create(train, session_id='exp_id', target='motion', prediction_window=50)` I get the following output: 
<img width="606" alt="screenshot 2018-08-11 18 28 13" src="https://user-images.githubusercontent.com/5934354/43996600-6aac9082-9d94-11e8-83f3-2ea9ce9f13e6.png">

Every project is unique so I don't expect anyone to debug this for me but any advice pointing me on what to read to learn how to resolve/understand that following error (`ToolkitError: Size of prediction probability vector(3) != number of classes(1).`) would be great.

**Comment by tetreault** [comment]
Second post: 

Hey guys! Sorry for hijacking this thread my mistake :) I understand if this should be in a new thread. 

For starters, I'm definitely using an older version, in my console I see `turicreate==4.3.2` which is just what pip installed. 

Overall my data i'm loading in to the SFrame has 4 sessions. 

The lines you asked me to run above result in the following: 

```
len(dataSFrame['exp_id'].unique()) # outputs 4
len(train['motion'].unique()) # outputs 3
len(test['motion'].unique()) # outputs 1
```

So therein lies the problem I suppose, if I get the error `ToolkitError: Size of prediction probability vector(3) != number of classes(1).` and i'm seeing 3 of the sessions are in train, but only 1 in test?

Thanks again for your time and response, as I mentioned I'm treading in very new water and am unfamiliar with ML as a whole but slowly trying to work on it!

**Comment by igiloh** [comment]
@tetreault - it seems that you dataset is too small. Please try to collect more data.  

It seems that you have collected a single session from each activity (motion) - you should collect a larger number of activities - taken at different times (maybe even from different users).  
Note that a session may include a person doing a variety of activities. You don't have to record data of someone only sitting, then only standing etc. It's ok if you do - but then try to have at least 4-5 different sessions of each activity.

You could try to artificially break down the data you already have to shorter sessions - but then you would probably encounter the original problem mentioned in #813 .

I'm closing the issue for now. Feel free to re-open if you have more questions or encounter further problems.

**Comment by tetreault** [comment]
Okay @igiloh I doubled back, grabbed all the HAPT data. I ran the steps in the [data prep page](https://apple.github.io/turicreate/docs/userguide/activity_classifier/data-preparation.html). Everything went smoothly. 

So now I'm using the code specified for creating the model (see [here](https://apple.github.io/turicreate/docs/userguide/activity_classifier/)). This data set now is the full HAPT data set, so much much larger than my 3.5k rows I put together myself. 

When I run the following line: `train, test = tc.activity_classifier.util.random_split_by_session(data, session_id='exp_id', fraction=0.8)`

I now receive the following warning: `The dataset has less than the minimum of 100 sessions required for train-validation split. Continuing without validation set` This results in an empty `test` set, so I cannot run `metrics = model.evaluate(test)`. 

Is this the problem you mentioned with `random_split_by_session` in your prior comment?

Thanks again for your time and any advice!

**Comment by tetreault** [comment]
interestingly enough, downgrading `turicreate` from v5.0b3 to v4.3.2 makes `random_split_by_session` work correctly. 

Running the following line in turicreate v4.3.2: 
`train, test = tc.activity_classifier.util.random_split_by_session(data, session_id='exp_id', fraction=0.8)`

Now properly gives me a training set with 580,082 rows, and a test set with 168,381 rows.

**Comment by igiloh** [comment]
Hi @tetreault,

Actually - it's the other way around 😄   
We've added a safeguard in v.50b3 that prevents from creating an automatic validation test if the dataset is too small. The reason, as I have mentioned above, is to try to prevent the case that both train and test sets contain only part of the labels (motions).   
I admit this decision is done rather crudely - we simply don't allow splitting if there's less than a certain amount of sessions (currently hard coded to 100).  

In addition - the safeguard was added mainly to prevent the `activity_classifier.create()` method function from creating an auto validation set. In no way it should block a user who is explicitly calling the `random_split_by_session()` API. 

I'm changing this issue to a feature request, so we will add:
1. As a first step - an enforce flag to `random_split_by_session()`, allowing the user to split his sessions on any dataset when using this API.
2. We should probably improve the decision mechanism for whether or not to create a validation set during `create()`. It should check the labels distribution of the train and validation sets (or some other method) - rather then a hard limit on the number of sessions.

**Comment by igiloh** [comment]
I have also created #969 - the error you got when you called `evaluate()` was a bit ambiguous. We can probably detect these cases right away and give a more informative error.

**Comment by tetreault** [comment]
You guys are awesome and on top of this stuff! So far my experience with turi and its maintainers is super positive. Thanks for the hard work guys!

**Comment by nickjong** [comment]
Labeling as a bug since the current experience seems wrong....

**Comment by nickjong** [comment]
We should definitely do step 1 of what Ilai suggested. Step 2 is really a separate matter, since this case is less about the validation_set='auto' and more about allowing users to invoke our validation-set creation code to produce test sets. #969 covers a separate case: the fact that we currently throw an exception if the test set doesn't happen to span all of the labels present in the training data.

With regard to step 2, note that most of our supervised toolkits currently just use the heuristic of using a validation set when the number of instances is above 100. I think the main question here is what is an "instance". A session feels like it's much bigger than an instance, since it may comprise several prediction windows, each of which has a separate (if not independent label). Given that our implementation already splits each session into segments of 20 prediction windows, we could in theory do the auto split if we have at least 100 of these 20-window segments. But that's nontrivial to compute accurately, and is probably best implemented on the C++ side as part of the chunking code. (Chunk first, then split.)

**Comment by igiloh** [comment]
@nickjong - I agree, the train-validation split probably should move to the C++ part. This will also save the need to calculate the number of sessions using SFrame operations - which is quite slow.

It does seem though that 100 sessions is a too high threshold for most datasets. It seems that most users collect much fewer data. Even the HAPT dataset (which we consider small, but may very well be a typical one) only has 42 sessions.  
For the time being - maybe we can add another threshold for `num_sessions` above which we do the split but verify that both sets include all possible labels. I.e.:
```python
if num_sessions < 30:
     #don't do split
if num_sessions > 30 and num_sessions <100:
     train, valid = split()
     if len(train[target].unique())) != len(valid[target].unique())):
           #give some warning\error
if num_sessions > 100:
      train, valid = split()
     # just split. Verifying will be probably quite slow
```

**Comment by nickjong** [comment]
Let's use #991 to track a fix for 6.0. I'm worried that we don't have time to adequately design and test a fix for 5.0, which we're releasing in just a couple of days. For 5.0, I'm pinning my hopes on #969: it shouldn't really be problematic if the validation set and training data span different numbers of labels.
